### PR TITLE
Clarification

### PR DESCRIPTION
**Pull request by Changron**
1. Only contestant can create new clarification/see clarify button
2. If user is not logged in or not contestant but send request, redirect to contestArchive

3.Fix typo in contest.models clarification realy_all --> reply_all

4.Should run "python manage.py makemigrations" and "python manage.py migrates" to apply changes in contest.models

### CODE REVIEW

File: contest/static/contest/js/function.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+copies or substantial portions of the Software.
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+function add_attribute(id, attribute, value) {
```
Comment (bruce3557) [inline contest/static/contest/js/function.js]: Is it possible to put this function in common library?
I think other guys may need this also.

Comment (bruce3557) [inline contest/static/contest/js/function.js]: And great code of this function, :+1:

Comment (Changron) [inline contest/static/contest/js/function.js]: I think so too, and thanks bruce

**Comment by nickelcheng** [comment]
I have no doubt with this PR. @CHIAHSINHOU please test this.

**Comment by CHIAHSINHOU** [comment]
1. 有點不懂 為什麼打在Description上的內容 顯示在 question 上
2. 按下 Clarify 後會出現 problem 的選單 應該只能出現 那場 contest 的題目 但現在是所有的題目 
3. problem 的選單 我覺得長度太長了 短一點比較好看
4. contest owner 跟 coowner 現在是不能回應嗎？

File: contest/static/contest/js/function.js (hunk 1 above)
Comment (Changron) [inline contest/static/contest/js/function.js]: please move to index folder

**Comment by Changron** [comment]
is this ok?
i got trouble moving the select up
but i think this is acceptable

**Comment by henryyang42** [comment]
Embrace both ``div`` to bootstrap class ``row`` and use ``col-xx-x`` to adjust the size

**Comment by Changron** [comment]
Thanks Henry, but I still couldn't make it.
1. question -> description
2. fixed
3. fixed
4. yes, functionality will be added later

**Comment by Changron** [comment]
merged.

**Comment by CHIAHSINHOU** [comment]
contest owner 跟 coowner 在不是Contestants時 沒有辦法 clarify 但他們應該不需要註冊比賽吧？

**Comment by Changron** [comment]
Should they need to be able to add new clarification?

**Comment by nickelcheng** [comment]
yeap, for announcement.

**Comment by Changron** [comment]
fix bug. add coowner and owner can add clarification now

**Comment by Changron** [comment]
Wait for review and test

**Comment by nickelcheng** [comment]
@CHIAHSINHOU please test again.

**Comment by CHIAHSINHOU** [comment]
你需要先merge 現在的dev噢 不然我不能測試

**Comment by Changron** [comment]
merged

File: contest/contest_info.py [hunk 2]
```diff
@@ -16,13 +16,13 @@
     LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
     OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
     SOFTWARE.
-    '''
+    '''
```
Comment (bruce3557) [inline contest/contest_info.py]: Remove the additional space at the end of line

File: contest/contest_info.py [hunk 3]
```diff
@@ -94,3 +96,20 @@ def get_scoreboard(contest):
         scoreboard.add_user(new_contestant)
 
     return scoreboard
+
+def get_clarifications(contest,user):
+    reply_all = Clarification.objects.filter(contest = contest,reply_all = True)
```
Comment (bruce3557) [inline contest/contest_info.py]: Add a space after ```,```

File: contest/contest_info.py [hunk 4]
```diff
@@ -94,3 +96,20 @@ def get_scoreboard(contest):
...
+    user_ask = Clarification.objects.filter(contest = contest,asker = user)
```
Comment (bruce3557) [inline contest/contest_info.py]: Ditto

**Comment by Changron** [comment]
space added

File: contest/templates/contest/clarification.html [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+{% endblock import_source %}
+<div class="modal fade" id="clarification" tabindex="-1" role="dialog">
+  <div class="modal-dialog">
+    <div class="modal-content">
+      <div class="modal-header">
+        <button type="button" class="close" data-dismiss="modal">
+          <span>&times;</span>
+          <span class="sr-only">Close</span>
+        </button>
+        <h4 class="modal-title" id="myModalLabel">Ask Question</h4>
```
Comment (bruce3557) [inline contest/templates/contest/clarification.html]: suggest that id change to ```clarificationModalLabel```

**Comment by bruce3557** [comment]
+1, wait for testing

**Comment by CHIAHSINHOU** [comment]
OK go to merge, but there are some work you need to do in next pull request.

**Pull request merged**